pkg/controller/opentelemetrycollector/reconcile_serviceaccount.go

```diff
@@ -0,0 +1,148 @@
+		// if this has been specified and is not empty, we manage the service account
+		svcs = append(svcs, serviceAccount(ctx))
+	}
+
+	// first, handle the create/update parts
+	if err := r.reconcileExpectedServiceAccounts(ctx, svcs); err != nil {
+		return fmt.Errorf("failed to reconcile the expected service accounts: %v", err)
+	}
+
+	// then, delete the extra objects
+	if err := r.deleteServiceAccounts(ctx, svcs); err != nil {
+		return fmt.Errorf("failed to reconcile the service accounts to be deleted: %v", err)
+	}
+
+	return nil
+}
+
+func serviceAccount(ctx context.Context) *corev1.ServiceAccount {
+	instance := ctx.Value(opentelemetry.ContextInstance).(*v1alpha1.OpenTelemetryCollector)
+	name := resourceName(instance.Name)
+
+	labels := commonLabels(ctx)
+	labels["app.kubernetes.io/name"] = name
+
+	return &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   instance.Namespace,
+			Labels:      labels,
+			Annotations: instance.Annotations,
+		},
+	}
+}
+
+func (r *ReconcileOpenTelemetryCollector) reconcileExpectedServiceAccounts(ctx context.Context, expected []*corev1.ServiceAccount) error {
+	logger := ctx.Value(opentelemetry.ContextLogger).(logr.Logger)
+	for _, obj := range expected {
+		desired := obj
+
+		// #nosec G104 (CWE-703): Errors unhandled.
+		r.setControllerReference(ctx, desired)
+
+		svcs := r.clientset.Kubernetes.CoreV1().ServiceAccounts(desired.Namespace)
+
+		existing, err := svcs.Get(desired.Name, metav1.GetOptions{})
+		if err != nil && errors.IsNotFound(err) {
+			if desired, err = svcs.Create(desired); err != nil {
+				return fmt.Errorf("failed to create: %v", err)
+			}
+
+			logger.WithValues("serviceAccount.name", desired.Name, "serviceAccount.namespace", desired.Namespace).V(2).Info("created")
+			continue
+		} else if err != nil {
+			return fmt.Errorf("failed to retrieve: %v", err)
+		}
+
+		// it exists already, merge the two if the end result isn't identical to the existing one
+		updated := existing.DeepCopy()
+		if updated.Annotations == nil {
+			updated.Annotations = map[string]string{}
+		}
+		if updated.Labels == nil {
+			updated.Labels = map[string]string{}
+		}
+		updated.ObjectMeta.OwnerReferences = desired.ObjectMeta.OwnerReferences
+
+		for k, v := range desired.ObjectMeta.Annotations {
+			updated.ObjectMeta.Annotations[k] = v
+		}
+		for k, v := range desired.ObjectMeta.Labels {
+			updated.ObjectMeta.Labels[k] = v
+		}
+
+		if updated, err = svcs.Update(updated); err != nil {
+			return fmt.Errorf("failed to apply changes to service account: %v", err)
+		}
+		logger.V(2).Info("applied", "serviceAccount.name", desired.Name, "serviceAccount.namespace", desired.Namespace)
+	}
+
+	return nil
+}
+
+func (r *ReconcileOpenTelemetryCollector) deleteServiceAccounts(ctx context.Context, expected []*corev1.ServiceAccount) error {
+	instance := ctx.Value(opentelemetry.ContextInstance).(*v1alpha1.OpenTelemetryCollector)
```

Comment:
Would it be better for non-public methods to just have the OpenTelemetryCollector instance, rather than keep accessing from context?

Reply:
In addition to the CR, the context contains a logger and possibly some trace context as well. We should probably consider moving the logger to the reconciler, but we'll still keep the ctx as the first parameters to most function calls, even if we pass the instance explicitly.